src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,462 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "actions-server.h"
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/CommandHandlerInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::Actions;
+using namespace chip::app::Clusters::Actions::Attributes;
+using namespace chip::Protocols::InteractionModel;
+
+namespace {
+static constexpr size_t kActionsDelegateTableSize =
+    MATTER_DM_ACTIONS_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT;
+static_assert(kActionsDelegateTableSize <= kEmberInvalidEndpointIndex, "Actions Delegate table size error");
+
+// TODO: We should not use global array, instead we can use one cluster instance per endpoint.
+Delegate * gDelegateTable[kActionsDelegateTableSize] = { nullptr };
+
+Delegate * GetDelegate(EndpointId aEndpoint)
+{
+    return (aEndpoint >= kActionsDelegateTableSize ? nullptr : gDelegateTable[aEndpoint]);
+}
+
+CHIP_ERROR IsValid(Delegate * delegate)
+{
+    if (delegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!!!");
```

Comment:
Should be logging the endpoint; otherwise it's not very actionable.

Reply:
Done